[FIXED JENKINS-50148] Handle SimpleBuildWrappers in editor

abayer opened the pull request:
# Description

See [JENKINS-50148](https://issues.jenkins-ci.org/browse/JENKINS-50148).

# Submitter checklist
- [x] Link to JIRA ticket in description, if appropriate.
- [x] Change is code complete and matches issue description
- [x] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

abayer commented:
@kshultzCB Any chance you could give this a quick manual test to see if things look right? Danke!

kzantow commented on pom.xml:
```diff
@@ -741,6 +741,13 @@
             </exclusions>
         </dependency>
 
+        <!-- used in blueocean-pipeline-editor test -->
```
NIT: since this is just used in the test, should it be defined there?

abayer replied:
Looked to me like dependencies are all defined in `<dependencyManagement>` here even if they're only used in one submodule, but eh, I don't care. =)

kshultzCB commented:
I'm hitting the same problem. Here's how I went about building this, in case something's gone wrong - but the latest commit looks like the right onw.

```
✔ ~/GitHub/blueocean-plugin [master|✔] 
15:09 $ git fetch jenkinsci  pull/1786/head:pr-1786
remote: Counting objects: 27, done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 27 (delta 11), reused 27 (delta 11), pack-reused 0
Unpacking objects: 100% (27/27), done.
From github.com:jenkinsci/blueocean-plugin
 * [new ref]             refs/pull/1786/head -> pr-1786
✔ ~/GitHub/blueocean-plugin [master|✔] 
15:10 $ git checkout pr-1786 
Switched to branch 'pr-1786'
✔ ~/GitHub/blueocean-plugin [pr-1786 L|✔] 
15:10 $ git hist
* 0f41447f3 (HEAD -> pr-1786) [FIXED JENKINS-50148] Handle SimpleBuildWrappers in editor

(snip)

```

Built with `mvn clean install -DcleanNode -DskipTests`, then run `mvn -f blueocean/pom.xml hpi:run`, and I've got a working local instance, with what the expected commit of Blue Ocean:

![image](https://user-images.githubusercontent.com/21689198/43860609-34ef2c30-9b22-11e8-8ad1-2d0994c0b3de.png)

Then I went through the Blue Ocean create flow. My internal Bitbucket server is already set up from previous work, and I've put a repo on it with the Jenkinsfile from the internal bug report. So all there is to it is to tell Blue Ocean to use that Bitbucket server, pick the repo, and off we go:

![image](https://user-images.githubusercontent.com/21689198/43860724-9b408efc-9b22-11e8-9ff7-ced32ca682fd.png)

The job gets stuck on "Waiting for next available executor", but that doesn't seem like it should matter for editing the Jenkinsfile. So, from the Activity tab, I click on the build of the `master` branch, and click the Edit pencil up near the top right of the screen:

![image](https://user-images.githubusercontent.com/21689198/43860908-2e9b7fd6-9b23-11e8-8310-aeb753fde031.png)

And I get a blank screen:

![image](https://user-images.githubusercontent.com/21689198/43860940-47a090ac-9b23-11e8-91ca-cb7b18a7e63e.png)

Network log shows the same problem:

```
Unhandled Rejection: "Error: No parameters for: withSonarQubeEnv
    at convertStepFromJson (http://127.0.0.1:8080/jenkins/adjuncts/2b879644/org/jenkins/ui/jsmodules/blueocean-pipeline-editor/jenkins-js-extension.js:72444:27)
    at convertStageFromJson (http://127.0.0.1:8080/jenkins/adjuncts/2b879644/org/jenkins/ui/jsmodules/blueocean-pipeline-editor/jenkins-js-extension.js:72395:28)
    at convertJsonToInternalModel (http://127.0.0.1:8080/jenkins/adjuncts/2b879644/org/jenkins/ui/jsmodules/blueocean-pipeline-editor/jenkins-js-extension.js:72351:28)
    at http://127.0.0.1:8080/jenkins/adjuncts/2b879644/org/jenkins/ui/jsmodules/blueocean-pipeline-editor/jenkins-js-extension.js:65882:96
    at http://127.0.0.1:8080/jenkins/adjuncts/2b879644/org/jenkins/ui/jsmodules/blueocean-pipeline-editor/jenkins-js-extension.js:72670:13
    at http://127.0.0.1:8080/jenkins/adjuncts/2b879644/org/jenkins/ui/jsmodules/blueocean-pipeline-editor/jenkins-js-extension.js:73492:17"

logUnhandledPromiseRejection	@	blueocean.js:58096

```

`Ctrl-S` trick doesn't bring anything up, either.

I'll attach a .har file for this too.

kshultzCB commented:
[I had to append `.zip` to the filename to get GitHub to take this.](https://github.com/jenkinsci/blueocean-plugin/files/2271812/No_parameters_for_withSonarQubeEnv.har.zip) Just remove it.

abayer commented:
Well, that’s a shame. Back to the drawing board!

kzantow commented:
@abayer there's some front-end stuff to be done to fix this, i should have a PR shortly. I think this fix is useful, however, also

abayer commented:
aaaah, I was testing `withSonarQubeEnv(installation: 'default')`, but not `withSonarQubeEnv('default')` - lemme figure that one out.

abayer commented:
@kshultzCB @kzantow - so I've updated the tests to actually verify the javascript side (i.e., I updated `StepMetadata.json` and added more tests), and added tests for `withSonarQubeEnv` specifically, most notably to handle the single required unnamed parameter case. I tested locally with `hpi:run`, and once I installed the Sonar scanner plugin, the editor loads. FYI.

kshultzCB commented:
I've rebuilt at `e0e5c6e38` and started again. Good news - it's working for me.

@abayer has a sample repo up for this issue, and the editor works fine for that one.

![image](https://user-images.githubusercontent.com/21689198/43919556-4db78402-9be4-11e8-8ae3-4b0697bd8fe6.png)

Works fine for the more complicated one from before, too:

![image](https://user-images.githubusercontent.com/21689198/43920452-ed627122-9be6-11e8-954d-4b19c856262c.png)

vivek commented:
Once we have green CI run, we can merge it. Nice one @abayer!
